zeth-contracts/contracts/MerkleTreeMiMC7.sol

```diff
@@ -8,43 +8,56 @@ import "./BaseMerkleTree.sol";
 import "./MiMC7.sol";
 
 contract MerkleTreeMiMC7 is BaseMerkleTree {
-  // Custom hash smart contract
-  MiMC7 public mimc7_hasher;
-
-  constructor(address hasher_address, uint treeDepth) BaseMerkleTree(treeDepth) public {
-    mimc7_hasher = MiMC7(hasher_address);
-  }
-
-  // Returns the current merkle tree
-  function getTree() public view returns (bytes32[] memory) {
-    uint nbNodes = 2**(depth + 1) - 1;
-    bytes32[] memory tmpTree = new bytes32[](nbNodes);
-    bytes32 left;
-    bytes32 right;
-    // Dump the leaves in the right indexes in the tree
-    for (uint i; i < nbLeaves; i++) {
-      tmpTree[(nbLeaves - 1) + i] = leaves[i];
-    }
-
-    // Compute the internal nodes of the merkle tree
-    for (uint i = nbLeaves - 2; i > 0; i--) {
-      left = tmpTree[2*i+1];
-      right = tmpTree[2*(i+1)];
 
-      // Seed is hardcoded and given by "clearmatics_mt_seed"
-      tmpTree[i] = mimc7_hasher.hash(left, right, "clearmatics_mt_seed");
+    constructor(uint treeDepth) BaseMerkleTree(treeDepth) public {
     }
 
-    // Compute the merkle root
-    left = tmpTree[1];
-    right = tmpTree[2];
-    tmpTree[0] = mimc7_hasher.hash(left, right, "clearmatics_mt_seed");
-
-    return tmpTree;
-  }
+    // Returns the current merkle tree
+    function getTree() public view returns (bytes32[] memory) {
+        uint nbNodes = 2**(depth + 1) - 1;
+        bytes32[] memory tmpTree = new bytes32[](nbNodes);
+        bytes32 left;
+        bytes32 right;
+        // Dump the leaves in the right indexes in the tree
+        for (uint i = 0; i < nbLeaves; i++) {
+            tmpTree[(nbLeaves - 1) + i] = leaves[i];
+        }
+
+        // Compute the internal nodes of the merkle tree
+        for (uint i = nbLeaves - 2; i > 0; i--) {
+            left = tmpTree[2*i+1];
+            right = tmpTree[2*(i+1)];
+
+            // Seed is hardcoded and given by "clearmatics_mt_seed"
+            tmpTree[i] = MiMC7.hash(left, right);
+        }
+
+        // Compute the merkle root
+        left = tmpTree[1];
+        right = tmpTree[2];
+        tmpTree[0] = MiMC7.hash(left, right);
+
+        return tmpTree;
+    }
 
-  // Returns the root of the merkle tree
-  function getRoot() public view returns(bytes32) {
-    return getTree()[0];
-  }
+    // Returns the root of the merkle tree
+    function getRoot() public view returns(bytes32) {
+        uint layerSize = nbLeaves / 2;
+        bytes32[nbLeaves/2] memory pad;
+
+        // Compute first layer from storage
+        for (uint i = 0 ; i < layerSize ; ++i) {
```

Comment:
` for (uint i = 0 ; i < layerSize ; ++i)` ->   `for (uint i = 0; i < layerSize; ++i)`, ie: remove spaces before `;`.

Note: There are inconsistencies in how for loops are written. Some are written like:
`for (uint i; i < layerSize; ++i)` (i.e. `i` is not explicitly initialized), and some are written as `for (uint i = 0; i < layerSize; ++i)` (explicit initialization).

I'd be in favor to use explicit initialization for readability and clarity (although Solidity initializes values to zero: https://solidity.readthedocs.io/en/v0.5.3/control-structures.html#default-value) if this does not hurt gas cost too much (it shouldn't except if I am missing something)

Reply:
> I'd be in favor to use explicit initialization for readability and clarity

I would tend to agree.  I'm sure the optimizer will remove any redundannt initialization.  I could be wrong, but I would be very surprised if it has any influence on gas usage.